components/suggestion/command_provider/command_provider.tsx

```diff
@@ -62,32 +67,84 @@ export class CommandSuggestion extends Suggestion {
     }
 }
 
+type Props = {
+    isInRHS: boolean;
+}
+
+export type Results = {
+    matchedPretext: string;
+    terms: string[];
+    items: AutocompleteSuggestion[];
+    component: React.ElementType;
+}
+
+type ResultsCallback = (results: Results) => void;
+
 export default class CommandProvider extends Provider {
-    constructor({isInRHS}) {
+    private isInRHS: boolean;
+
+    private parser: AppCommandParser;
+    store: Store;
+
+    constructor(props: Props, store?: Store) {
         super();
 
-        this.isInRHS = isInRHS;
+        this.store = globalStore;
+        if (store && store.getState && store.dispatch) {
+            this.store = store;
+        }
+
+        this.isInRHS = props.isInRHS;
+        let rootId;
+        if (this.isInRHS) {
+            const selectedPost = getSelectedPost(this.store.getState()) as Post;
+            if (selectedPost) {
+                rootId = selectedPost.root_id ? selectedPost.root_id : selectedPost.id;
+            }
+        }
+
+        this.parser = new AppCommandParser(this.store.dispatch, this.store.getState, rootId);
     }
 
-    handlePretextChanged(pretext, resultCallback) {
+    handlePretextChanged(pretext: string, resultCallback: ResultsCallback) {
         if (!pretext.startsWith('/')) {
             return false;
         }
+
+        const command = pretext.toLowerCase();
+        if (this.parser.isAppCommand(command)) {
+            this.parser.getAppCommandSuggestions(command).then((matches) => {
+                const terms = matches.map((suggestion) => suggestion.complete);
+                resultCallback({
+                    matchedPretext: command,
+                    terms,
+                    items: matches,
+                    component: CommandSuggestion,
+                });
+            });
+            return true;
+        }
+
         if (UserAgent.isMobile()) {
-            return this.handleMobile(pretext, resultCallback);
+            this.handleMobile(pretext, resultCallback);
+        } else {
+            this.handleWebapp(pretext, resultCallback);
         }
-        return this.handleWebapp(pretext, resultCallback);
+        return true;
     }
 
-    handleCompleteWord(term, pretext, callback) {
+    handleCompleteWord(term: string, pretext: string, callback: (s: string)=>void) {
         callback(term + ' ');
     }
 
-    handleMobile(pretext, resultCallback) {
+    handleMobile(pretext: string, resultCallback: ResultsCallback) {
         const command = pretext.toLowerCase();
         Client4.getCommandsList(getCurrentTeamId(store.getState())).then(
             (data) => {
-                let matches = [];
+                let matches: AutocompleteSuggestion[] = [];
+                const appCommandSuggestions = this.parser.getAppSuggestionsForBindings(pretext);
```

Comment:
App suggestion for bidings, or app suggestion for commands? Is there any other type of binding where we will get suggestions like this? If so, how do we specify the type of binding? Or there is no need for such specification?